### PR TITLE
[WIP] Examples for time frequency analysis on SourceEstimates

### PR DESCRIPTION
**Pull request by DiGyt**
#### Reference issue
Part of GSoC: TFR in Source Space Project (See: Issue #6290 ).

#### What does this implement/fix?
Add examples on how to compute time frequency analysis in source space,
using functions from mne.time_frequency and SourceEstimates.

#### Additional information
This branch is forked from DiGyt:plot_source_tfr (PR #6672),
and therefore should be merged after PR #6672 (and PR #6629).

### CODE REVIEW

File: mne/minimum_norm/inverse.py [hunk 1]
```diff
@@ -1071,12 +1071,19 @@ def apply_inverse_raw(raw, inverse_operator, lambda2, method="dSPM",
...
     _check_option('method', method, INVERSE_METHODS)
     _check_ori(pick_ori, inverse_operator['source_ori'])
     _check_ch_names(inverse_operator, epochs.info)
 
+    is_free_ori = not (is_fixed_orient(inverse_operator) or
+                       pick_ori == 'normal')
+
+    if delayed and is_free_ori and pick_ori != "vector":
+        raise ValueError("delayed must be False for free orientations other "
+                         "than pick_ori='vector'.")
```
Comment (agramfort) [inline mne/minimum_norm/inverse.py]: I think it should be `pick_ori='vector' or 'normal'`

File: examples/time_frequency/plot_source_tfr_power_and_itc.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+from mne.datasets import sample
+from mne.minimum_norm import read_inverse_operator, apply_inverse_epochs
+from mne.time_frequency import tfr_multitaper
+
+print(__doc__)
+
+###############################################################################
+# Set parameters
+data_path = sample.data_path()
+raw_fname = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw.fif')
```
Comment (agramfort) [inline examples/time_frequency/plot_source_tfr_power_and_itc.py]: I would not use sample for such an example but rather somato. It contains real induced power.

File: examples/time_frequency/plot_source_tfr_power_and_itc.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+fname_inv = op.join(data_path, 'MEG', 'sample',
+                    'sample_audvis-meg-oct-6-meg-fixed-inv.fif')
```
Comment (agramfort) [inline examples/time_frequency/plot_source_tfr_power_and_itc.py]: don't use a fixed ori inv if you can and rather use pick_ori='normal' below. It's more consistent with common MNE usage.

File: examples/time_frequency/plot_source_tfr_power_and_itc.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+# return_generator=True will reduce memory load
+# delayed=True will reduce time needed for computation
+stcs = apply_inverse_epochs(epochs, inverse_operator, lambda2=1./9.,
+                            method="dSPM", pick_ori="vector", prepared=False,
+                            return_generator=True, delayed=True)
+
+# calculate a time-frequency transform in source space
+freqs = np.arange(8, 18, 2)
+pow, _ = tfr_multitaper(stcs, freqs=freqs, n_cycles=2, use_fft=True,
+                              average=True, return_itc=True)
```
Comment (agramfort) [inline examples/time_frequency/plot_source_tfr_power_and_itc.py]: set `return_itc=False` and no need for `, _`

File: mne/source_tfr.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
+
```
Comment (agramfort) [inline mne/source_tfr.py]: why not putting this file in `mne.time_frequency` ?

File: mne/source_tfr.py [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+        write_hdf5(fname,
+                   dict(vertices=self._vertices, data=self.data,
+                        tmin=self.tmin, tstep=self.tstep,
+                        subject=self.subject, src_type=self._src_type),
+                   title='mnepython', overwrite=True)
+
+    @property
+    def sfreq(self):
+        """Sample rate of the data."""
+        return 1. / self.tstep
```
Comment (agramfort) [inline mne/source_tfr.py]: we should have a base or mixin class to avoid duplicating such properties

File: mne/source_tfr.py [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+                           '.tmin, .tstep or .data changes.')
+
+    def _update_times(self):
+        """Update the times attribute after changing tmin, tmax, or tstep."""
+        self._times = self.tmin + (self.tstep * np.arange(self.shape[-1]))
+        self._times.flags.writeable = False
+
+    def copy(self):
+        """Return copy of SourceTFR instance."""
+        return copy.deepcopy(self)
```
Comment (agramfort) [inline mne/source_tfr.py]: the methods / properties above should be factorized using _BaseSourceEstimate class

File: mne/tests/test_source_tfr.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+import os.path as op
+
+import numpy as np
+from numpy.testing import (assert_array_equal,
+                           assert_allclose, assert_equal)
+import pytest
+from mne.utils import _TempDir, requires_h5py, run_tests_if_main
+from mne.source_tfr import SourceTFR
+
+np.random.seed(seed=23)
```
Comment (agramfort) [inline mne/tests/test_source_tfr.py]: never affect the global seed. Use a random_state

File: mne/tests/test_source_tfr.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
+from copy import deepcopy
```
Comment (agramfort) [inline mne/tests/test_source_tfr.py]: author + license missing

File: mne/time_frequency/_stockwell.py [hunk 10]
```diff
@@ -248,21 +250,51 @@ def tfr_stockwell(inst, fmin=None, fmax=None, n_fft=None,
...
+
+    if isinstance(inst, list) or isgenerator(inst):
+        for idx, obj in enumerate(inst):
+            if idx == 0:
+                data = _get_data(obj, return_itc=False)  # don't provoke errors
+                inst = obj  # overwrite this to determine the object type later
+            else:
+                # data is simply concatenated. This should be changed later.
+                data = np.concatenate((data, _get_data(obj, return_itc=False)),
+                                      axis=0)
```
Comment (agramfort) [inline mne/time_frequency/_stockwell.py]: this is going to blow up the memory no?

Comment (DiGyt) [reply mne/time_frequency/_stockwell.py]: Yes! that's the next things I'll work on!

Comment (DiGyt) [reply mne/time_frequency/_stockwell.py]: it's changed now

File: mne/minimum_norm/inverse.py (hunk 1 above)
Comment (DiGyt) [reply mne/minimum_norm/inverse.py]: Changed this in PR #6609 , so we can rebase it when it's merged

**Comment by DiGyt** [comment]
Closed: See Issue #6290

**Pull request closed without merging**